sycl/doc/extensions/experimental/sycl_ext_oneapi_virtual_mem.asciidoc

```diff
@@ -0,0 +1,363 @@
+
+To represent the underlying physical device memory a virtual address is mapped
+to, the `physical_mem` class is added. This new class is defined as:
+
+```c++
+namespace sycl::ext::oneapi::experimental {
+
+enum class address_access_mode : /*unspecified*/ {
+  none,
+  read,
+  read_write
+};
+
+class physical_mem {
+public:
+  physical_mem(const device &syclDevice, const context &syclContext, size_t numBytes);
+  physical_mem(const queue &syclQueue, size_t numBytes);
+
+  /* -- common interface members -- */
+
+  void *map(uintptr_t ptr, size_t numBytes, size_t offset = 0) const;
+  void *map(uintptr_t ptr, size_t numBytes, address_access_mode mode, size_t offset = 0) const;
+
+  context get_context() const;
+  device get_device() const;
+
+  size_t size() const noexcept;
+};
+
+} // namespace sycl::ext::oneapi::experimental
+```
+
+`physical_mem` has common reference semantics, as described in
+{crs}[section 4.5.2. Common reference semantics].
+
+[frame="topbot",options="header,footer"]
+|============================
+|Member function |Description
+
+|`physical_mem(const device &syclDevice, const context &syclContext, size_t numBytes)` |
+Constructs a `physical_mem` instance using the `syclDevice` provided. This
+device must either be contained by `syclContext` or it must be a descendent
+device of some device that is contained by that context, otherwise this function
+throws a synchronous exception with the `errc::invalid` error code.
+
+This will allocate `numBytes` of physical memory on the device. `numBytes` must
+be a multiple of the granularity, as returned by a call to
+`get_mem_granularity`.
+
+If `syclDevice` does not have `aspect::ext_oneapi_virtual_mem` the call throws
+an exception with `errc::feature_not_supported`.
+
+|`physical_mem(const queue &syclQueue, size_t numBytes)` |
+Same as `physical_mem(syclQueue.get_device(), syclQueue.get_context, numBytes)`.
+
+|`void *map(uintptr_t ptr, size_t numBytes, size_t offset = 0)` |
+Same as `map(ptr, numBytes, address_access_mode::none, offset)`.
+
+|`void *map(uintptr_t ptr, size_t numBytes, address_access_mode mode, size_t offset = 0)` |
+Maps a virtual memory range, specified by `ptr` and `numBytes`, to the physical
+memory corresponding to this instance of `physical_mem`, starting at an offset
+of `offset` bytes.
+
+If `mode` is `address_access_mode::read` or `address_access_mode::read_write`
+the returned pointer is accessible after the call as read-only or read-write
+respectively. Otherwise, it is considered inaccessible and accessing it will
+result in undefined behavior.
+
+The returned pointer is equivalent to `reinterpret_cast<void *>(ptr)`.
+
+Writing to any address in the virtual memory range with access mode set to
+`access_mode::read` results in undefined behavior.
+
+An accessible pointer behaves the same as a pointer to device USM memory and can
+be used in place of a device USM pointer in any interface accepting one.
```

Comment:
> Can you map a new virtual address over physical memory that is already mapped to a different address?

I see nothing in both CUDA and L0 disallowing this.

> If a virtual memory address is mapped to some physical memory, can that virtual memory also be mapped to some different physical memory?

This is definitely not allowed. The suggested wording has been added. 👍

Reply:
>> Can you map a new virtual address over physical memory that is already mapped to a different address?
>
> I see nothing in both CUDA and L0 disallowing this.

That's interesting.  What happens in this case?  Is the physical memory automatically unmapped from the old virtual address range?

Reply:
I would assume it would be like having two pointers to the same memory location, but I don't know for certain.

Reply:
Do we have any tests for this scenario?  It seems like a weird case, so I think we should have some tests if we are going to say that it is supported.  For example, the test should try writing a value to one address and then reading from the other.  Maybe something like:

```
int *p1 = /* some mapped address */;
int *p2 = /* a different address that points to the same physical memory */;

*p1 = VALUE1;
*p2 = VALUE2;
assert(*p1 == VALUE2);
```

I would actually be surprised if this works because:

* The compiler would need to know that `p1` and `p2` alias each other.
* The hardware cache would need to know that `p1` and `p2` alias each other.